File: app/runner/test-limits.mjs
// Keep defaults and execution budgets aligned with backend/services/test_limits.py.
export const HARD_MAX_TEST_STEPS = 10000;
export function validateMaxSteps(value) {
  if (!Number.isInteger(value) || value < 2 || value > HARD_MAX_TEST_STEPS) {
    throw Error('测试协议错误：测试步骤上限必须是 2–10000 的整数');
  }
  return value;
}
export function configuredMaxSteps(raw = process.env.ATOMFORGE_MAX_TEST_STEPS ?? '1000') {
  if (typeof raw !== 'string' || !/^[0-9]+$/.test(raw)) throw Error('ATOMFORGE_MAX_TEST_STEPS 必须是 2–10000 的整数');
  return validateMaxSteps(Number(raw));
}
export const MAX_TEST_STEPS = configuredMaxSteps();
export function testTimeoutMs(stepCount) {
  return Math.max(60, 30 + 3 * Math.min(Math.max(stepCount, 0), HARD_MAX_TEST_STEPS)) * 1000;
}
export function jobTimeoutMs(stepCount) { return testTimeoutMs(stepCount) + 30000; }

File: app/backend/services/test_limits.py
"""Test-plan defaults and bounded execution budgets, mirrored by the runner."""
import os

HARD_MAX_TEST_STEPS = 10000


def validate_max_steps(value):
    if type(value) is not int or not 2 <= value <= HARD_MAX_TEST_STEPS:
        raise ValueError(f'测试步骤上限必须是 2–{HARD_MAX_TEST_STEPS} 的整数')
    return value


def configured_max_steps():
    raw = os.getenv('ATOMFORGE_MAX_TEST_STEPS', '1000')
    if not raw.isascii() or not raw.isdecimal():
        raise ValueError('ATOMFORGE_MAX_TEST_STEPS 必须是 2–10000 的整数')
    return validate_max_steps(int(raw))


MAX_TEST_STEPS = configured_max_steps()


def test_timeout_seconds(step_count):
    # Allow each action its 2.5s timeout plus overhead, with time for mounting.
    return max(60, 30 + 3 * min(max(step_count, 0), HARD_MAX_TEST_STEPS))


def runner_request_timeout(step_count):
    # Browser deadline < runner watchdog (+30s) < HTTP read deadline (+60s).
    return test_timeout_seconds(step_count) + 60
